src/model/reference.ts

```diff
@@ -56,39 +58,70 @@ export class ScannedReference<K extends keyof FeatureKindMap> extends
+    let feature: undefined|FeatureKindMap[DK];
     const warnings = [...this.warnings];
-    if (this.sourceRange) {
-      if (features.size === 0) {
-        let message = `Could not resolve reference to ${this.kind}`;
-        if (kind === 'behavior') {
-          message += `. Is it annotated with @polymerBehavior?`;
+
+    const binding = this.astPath.scope.getBinding(this.identifier);
+    if (binding !== undefined) {
+      const result = resolveBinding(binding, document, kind);
+      if (result.successful) {
+        feature = result.value;
+      }
+    }
+    if (feature === undefined) {
+      // We didn't find it by doing principled scope-based analysis. Let's try
+      // looking it up in our big global map!
+      const features = document.getFeatures(
+          {imported: true, externalPackages: true, kind, id: this.identifier});
+      if (this.sourceRange) {
+        if (features.size === 0) {
+          let message = `Could not resolve reference to ${this.kind}`;
+          if (kind === 'behavior') {
+            message += `. Is it annotated with @polymerBehavior?`;
+          }
+          warnings.push(new Warning({
+            code: 'could-not-resolve-reference',
+            sourceRange: this.sourceRange,
+            message,
+            parsedDocument: document.parsedDocument,
+            severity: Severity.WARNING
+          }));
+        } else if (features.size > 1) {
+          warnings.push(new Warning({
+            code: 'multiple-global-declarations',
+            sourceRange: this.sourceRange,
+            message: `Multiple global declarations of ${
+                this.kind} with identifier ${this.identifier}`,
+            parsedDocument: document.parsedDocument,
+            severity: Severity.WARNING
+          }));
         }
-        warnings.push(new Warning({
-          code: 'could-not-resolve-reference',
-          sourceRange: this.sourceRange,
-          message,
-          parsedDocument: document.parsedDocument,
-          severity: Severity.WARNING
-        }));
-      } else if (features.size > 1) {
-        warnings.push(new Warning({
-          code: 'multiple-global-declarations',
-          sourceRange: this.sourceRange,
-          message: `Multiple global declarations of ${
-              this.kind} with identifier ${this.identifier}`,
-          parsedDocument: document.parsedDocument,
-          severity: Severity.WARNING
-        }));
       }
+      [feature] = features;
     }
-    let feature: undefined|FeatureKindMap[DK];
-    [feature] = features;
     return new Reference<FeatureKindMap[K]>(this, feature, warnings);
   }
 }
 
+function resolveBinding<K extends keyof FeatureKindMap>(
+    binding: Binding, document: Document, kind: K):
+    Result<FeatureKindMap[K], Warning|undefined> {
+  const statement = esutil.getCanonicalStatement(binding.path);
+  if (!statement) {
+    return {successful: false, error: undefined};
```

Comment:
Add some kind of an error? In the above code it'll be ignored, but in a case where a caller does expect this to be resolved, they might want to know why it failed.

Reply:
Did the hookup here, filed https://github.com/Polymer/polymer-analyzer/issues/919 for actually emitting them, which might mean passing more context down.